<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Introduction: a pipeline to analyse time-course gene expression data</title>






<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction: a pipeline to analyse time-course gene expression data</h1>



<div id="abstract" class="section level1">
<h1>Abstract</h1>
<p>The phenotypic diversity of cells is governed by a complex equilibrium between their genetic identity and their environmental interactions: Understanding the dynamics of gene expression is a fundamental question of biology. However, analysing time-course transcriptomic data raises unique challenging statistical and computational questions, requiring the development of novel methods and software. Using as case study time-course transcriptomics data from mice exposed to different strains of influenza, ihis workflows provides a step-by-step tutorial of the methodology used to analyse time-course data: (1) normalization of the micro-array dataset; (2) differential expression analysis using functional data analysis; (3) clustering fo time-course data; (4) interpreting clusters with GO term and KEGG pathway enrichment analysis.</p>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Gene expression studies provide simultaneous quantification of the level of mRNA from all genes in a sample. High-throughput studies of gene expression have a long history, starting with microarray technologies in the 1990s through to single-cell technologies. While many expression studies are designed to compare the gene expression in distinct groups, there is also a long history of time-course expression studies, where the the gene expression is compared across time by measuring mRNA levels from different samples across time.<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> Such time course studies can vary from measuring a few distinct time points, to sampling ten to twenty time points. Many longer time series are particularly interested in investigating development over time. More recently, single-cell studies track single cells through their development, and a single cell is measured at a particular moment in its developmental progression – a value that is not know but estimated from the data as its “pseudo-time.”</p>
<p>While there are many methods that have been proposed for discrete aspects of time course data, the entire workflow for analysis of such data remains difficult, particularly for long, developmental time series. Most methods proposed for time course data are concerned with detecting genes that are changing over time (differential expression analysis), examples being <code>edge</code> <span class="citation">(Storey et al. 2005)</span>, functional component analysis based models <span class="citation">(Wu and Wu 2013)</span>, time-course permutation tests <span class="citation">(Park et al. 2003)</span>, XXX <span class="citation">(Wenguang and Zhi 2011)</span> <strong>TODO: Other methods to mention: impulse models, what else? Citations!</strong>. However, with long time course datasets, particularly in developmental systems, a massive number of genes will show <em>some</em> change. For example, in the system, over % of genes are shown to be changing over time. The task in these settings is often not to detect changes in genes, but to categorize into biologically interpretable patterns the vast number of changes discovered.</p>
<p>We present here a workflow for such an analysis that consists of 4 main parts (Figure ):</p>
<ul>
<li>Quality control and normalization;</li>
<li>Identification of genes that are differentially expressed;</li>
<li>Clustering of genes into distinct temporal patterns;</li>
<li>Biological interpretation of the clusters.</li>
</ul>
<p>This workflow represents an integration of both novel implementations of previously established methods and new methodologies for the settings of developmental time series. We provide the various steps of the workflow as functions in a R package <strong>FIXME</strong>.</p>
</div>
<div id="analysis-of-the-dynamical-response-of-mouse-lung-tissue-to-influenza" class="section level1">
<h1>Analysis of the dynamical response of mouse lung tissue to influenza</h1>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>This workflow is illustrated using data from a micro-array time-course experiment, exposing mice to three different strains of influenza, and collecting lung tissue during 14 time-points after infection (0, 3, 6, 9, 12, 18, 24, 30, 36, 48, 60 hours, then 3, 5, and 7 days later) <span class="citation">(Shoemaker 2015)</span>. The three strains of influenza used in the study are (1) a low pathogenicity seasonal H1N1 influenza virus (A/Kawasaki/UTK4/2009 [H1N1]), a mildly pathogenic virus from the 2009 pandemic season (A/California/04/2009 [H1N1]), and a highly pathogenic H5N1 avian influenza virus (A/Vietnam/1203/2004 [H5N1]. Mice were injected with 10^5 PFU of each virus. An adidtional 42 mice were injected with a lower dose of the Vietnam avian influenza virus (10^3 PFU).</p>
<p>The authors find that</p>
<p><strong>XXX Add summary of the key findings of the the paper</strong></p>
</div>
<div id="references" class="section level2 unnumbered">
<h2>References</h2>
<div id="refs" class="references">
<div id="ref-park:statistical">
<p>Park, Taesung, Dong-Hyun Yoo, Jun-Ik Ahn, Seung Yeoun Lee, Seungmook Lee, Sung-Gon Yi, and Yong-Sung Lee. 2003. “Statistical tests for identifying differentially expressed genes in time-course microarray experiments.” <em>Bioinformatics</em> 19 (6): 694–703. <a href="https://doi.org/10.1093/bioinformatics/btg068" class="uri">https://doi.org/10.1093/bioinformatics/btg068</a>.</p>
</div>
<div id="ref-shoemaker:ultrasensitive">
<p>Shoemaker, S. AND Eisfeld, J. E. AND Fukuyama. 2015. “An Ultrasensitive Mechanism Regulates Influenza Virus-Induced Inflammation.” <em>PLoS Pathogens</em> 11 (6): 1–25.</p>
</div>
<div id="ref-storey:significance">
<p>Storey, J. D., W. Xiao, J. T. Leek, R. G. Tompkins, and R. W. Davis. 2005. “Significance analysis of time course microarray experiments.” <em>Proc. Natl. Acad. Sci. U.S.A.</em> 102 (36): 12837–42.</p>
</div>
<div id="ref-sun:multiple">
<p>Wenguang, Sun, and Wei Zhi. 2011. “Multiple Testing for Pattern Identification, with Applications to Microarray Time-Course Experiments.” <em>Journal of the American Statistical Association</em> 106 (493). Taylor &amp; Francis: 73–88. <a href="https://doi.org/10.1198/jasa.2011.ap09587" class="uri">https://doi.org/10.1198/jasa.2011.ap09587</a>.</p>
</div>
<div id="ref-wu:more">
<p>Wu, Shuang, and Hulin Wu. 2013. “More Powerful Significant Testing for Time Course Gene Expression Data Using Functional Principal Component Analysis Approaches.” <em>BMC Bioinformatics</em> 14 (1): 6. <a href="https://doi.org/10.1186/1471-2105-14-6" class="uri">https://doi.org/10.1186/1471-2105-14-6</a>.</p>
</div>
</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>Because the collection of the mRNA is often destructive, samples at different time points are generally from different biological samples; longitudinal studies, for example tracking the same subject over time, are certainly possible in certain settings, but not directly considered here.<a href="#fnref1" class="footnote-back">↩</a></p></li>
</ol>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
