<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Downstream analysis: A pipeline to analyse time-course gene expression data</title>

<script>$(document).ready(function(){
    if (typeof $('[data-toggle="tooltip"]').tooltip === 'function') {
        $('[data-toggle="tooltip"]').tooltip();
    }
    if ($('[data-toggle="popover"]').popover === 'function') {
        $('[data-toggle="popover"]').popover();
    }
});
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>

</head>

<body>




<h1 class="title toc-ignore">Downstream analysis: A pipeline to analyse time-course gene expression data</h1>



<pre><code>## Warning: no function found corresponding to methods exports from 'XVector'
## for: 'concatenateObjects'</code></pre>
<div id="downstream-analysis-of-clusters." class="section level2">
<h2>Downstream analysis of clusters.</h2>
<p>Once good clusters are obtained, the next step is to leverage the clustering to ease interpretation. Classic enrichment analysis step can then be performed in the gene set defined in each cluster: KEGG pathway enrichment analysis, GO term enrichment analysis, motif enrichment analysis, etc.</p>
<p>First, let us clean up the gene obtained and only select the genes we are going to use in the enrichment analysis. One can either use the whole set of genes, only the set of differentially expressed genes in each cluster, or a subset of genes that fit well to a cluster (based on some criterion).</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">gene_names =<span class="st"> </span><span class="kw">row.names</span>(labels)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">gene_names =<span class="st"> </span>gene_names[<span class="op">!</span><span class="kw">is.na</span>(labels)]</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">labels =<span class="st"> </span>labels[<span class="op">!</span><span class="kw">is.na</span>(labels)]</a></code></pre></div>
<div id="finding-enriched-pathways-using-biomart-and-keggprofile" class="section level3">
<h3>Finding enriched pathways using <code>biomaRt</code> and <code>KEGGprofile</code></h3>
<p>Let us first tackle the case of pathway enrichment analysis. We will leverage the packages <code>biomaRt</code> and <code>KEGGprofile</code> for this step. <code>KEGGprofile</code> is a package that easily allows to perform pathway enrichment analysis on a set of genes labeled with the ensembl annotation. We thus need to convert the gene names into the appropriate format. This is where <code>biomaRt</code> comes in handy: it enables easy conversion from one gene annotation to another. Here, we will use it to convert the gene names from the Refseq annotation to the ensembl one.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">library</span>(biomaRt)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">ensembl =<span class="st"> </span><span class="kw">useMart</span>(<span class="st">&quot;ensembl&quot;</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">ensembl =<span class="st"> </span><span class="kw">useDataset</span>(<span class="st">&quot;mmusculus_gene_ensembl&quot;</span>, <span class="dt">mart=</span>ensembl)</a></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">clusters =<span class="st"> </span><span class="kw">sort</span>(<span class="kw">as.vector</span>(<span class="kw">unlist</span>(<span class="kw">unique</span>(labels))))</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">clusters =<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">all_pathways =<span class="st"> </span><span class="kw">data.frame</span>()</a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="cf">for</span>(cluster <span class="cf">in</span> clusters){</a>
<a class="sourceLine" id="cb4-5" data-line-number="5">    genes =<span class="st"> </span>gene_names[labels <span class="op">==</span><span class="st"> </span>cluster]</a>
<a class="sourceLine" id="cb4-6" data-line-number="6"></a>
<a class="sourceLine" id="cb4-7" data-line-number="7">    <span class="co"># convert gene names</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8">    genes =<span class="st"> </span><span class="kw">getBM</span>(<span class="dt">attributes=</span><span class="kw">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;entrezgene&quot;</span>),</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">          <span class="dt">filters=</span><span class="st">&quot;refseq_mrna&quot;</span>, <span class="dt">values=</span>genes,</a>
<a class="sourceLine" id="cb4-10" data-line-number="10">          <span class="dt">mart=</span>ensembl)[<span class="st">&quot;entrezgene&quot;</span>]</a>
<a class="sourceLine" id="cb4-11" data-line-number="11">    genes =<span class="st"> </span><span class="kw">as.vector</span>(<span class="kw">unlist</span>(genes))</a>
<a class="sourceLine" id="cb4-12" data-line-number="12">    pathways =<span class="st"> </span>KEGGprofile<span class="op">::</span><span class="kw">find_enriched_pathway</span>(</a>
<a class="sourceLine" id="cb4-13" data-line-number="13">    genes, <span class="dt">species=</span><span class="st">&quot;mmu&quot;</span>,</a>
<a class="sourceLine" id="cb4-14" data-line-number="14">    <span class="dt">download_latest=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-15" data-line-number="15"></a>
<a class="sourceLine" id="cb4-16" data-line-number="16">    <span class="cf">if</span>((<span class="kw">dim</span>(all_pathways)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&amp;</span><span class="st"> </span>(<span class="kw">dim</span>(pathways<span class="op">$</span>stastic)[<span class="dv">1</span>] <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>)){</a>
<a class="sourceLine" id="cb4-17" data-line-number="17">    <span class="kw">print</span>(<span class="st">&quot;Initialize&quot;</span>)</a>
<a class="sourceLine" id="cb4-18" data-line-number="18">    pathways<span class="op">$</span>stastic[<span class="st">&quot;Cluster&quot;</span>] =<span class="st"> </span>cluster</a>
<a class="sourceLine" id="cb4-19" data-line-number="19">    all_pathways =<span class="st"> </span>pathways<span class="op">$</span>stastic</a>
<a class="sourceLine" id="cb4-20" data-line-number="20">    }<span class="cf">else</span> <span class="cf">if</span>(<span class="kw">dim</span>(pathways<span class="op">$</span>stastic)[<span class="dv">1</span>] <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>) {</a>
<a class="sourceLine" id="cb4-21" data-line-number="21">    <span class="kw">print</span>(<span class="st">&quot;Merge&quot;</span>)</a>
<a class="sourceLine" id="cb4-22" data-line-number="22">    pathways<span class="op">$</span>stastic[<span class="st">&quot;Cluster&quot;</span>] =<span class="st"> </span>cluster</a>
<a class="sourceLine" id="cb4-23" data-line-number="23">    all_pathways =<span class="st"> </span><span class="kw">merge</span>(all_pathways, pathways<span class="op">$</span>stastic, <span class="dt">all=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-24" data-line-number="24">    }</a>
<a class="sourceLine" id="cb4-25" data-line-number="25">}</a></code></pre></div>
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Pathway
</th>
<th style="text-align:right;">
Percentage
</th>
<th style="text-align:right;">
Adj. p-value
</th>
<th style="text-align:right;">
Cluster
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
Oxidative phosphorylation
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0.0156
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
12
</td>
<td style="text-align:left;">
Thermogenesis
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0156
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
11
</td>
<td style="text-align:left;">
Ribosome
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0414
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
Metabolic pathways
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0.0427
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
Huntington disease
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0493
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
Non-alcoholic fatty liver disease (NAFLD)
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0493
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
Parkinson disease
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0493
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
10
</td>
<td style="text-align:left;">
Retrograde endocannabinoid signaling
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0493
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
Olfactory transduction
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0.0000
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
Neuroactive ligand-receptor interaction
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0.0049
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
ECM-receptor interaction
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
0.0075
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:left;">
Human papillomavirus infection
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0.0075
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
13
</td>
<td style="text-align:left;">
Tight junction
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0.0075
</td>
<td style="text-align:right;">
3
</td>
</tr>
</tbody>
</table>
</div>
<div id="finding-enriched-go-terms" class="section level3">
<h3>Finding enriched GO terms</h3>
<p>To find GO terms, we use <code>biomaRt</code> to find the mapping between GO terms and gene mapping. The GO enrichment library <code>topGO</code> expects the GO term to gene mapping to be a list where each item is a mapping between a gene name and a GO term ID vector.</p>
<pre class="text"><code>$NM_199153
[1] &quot;GO:0016020&quot; &quot;GO:0016021&quot; &quot;GO:0007186&quot; &quot;GO:0004930&quot; &quot;GO:0007165&quot;
[6] &quot;GO:0050896&quot; &quot;GO:0050909&quot;

$NM_201361
 [1] &quot;GO:0016020&quot; &quot;GO:0016021&quot; &quot;GO:0003674&quot; &quot;GO:0008150&quot; &quot;GO:0005794&quot;
 [6] &quot;GO:0005829&quot; &quot;GO:0005737&quot; &quot;GO:0005856&quot; &quot;GO:0005874&quot; &quot;GO:0005739&quot;
[11] &quot;GO:0005819&quot; &quot;GO:0000922&quot; &quot;GO:0072686&quot;</code></pre>
<p><code>biomaRt</code> queries results a matrix with two named columns of gene names and GO term ID. <strong>FIXME</strong> matrix or data.frame??</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># </span><span class="al">FIXME</span><span class="co"> move this to the package</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw">source</span>(<span class="st">&quot;GO_terms_analysis.R&quot;</span>)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">genes =<span class="st"> </span><span class="kw">getBM</span>(<span class="dt">attributes=</span><span class="kw">c</span>(<span class="st">&quot;go_id&quot;</span>, <span class="st">&quot;refseq_mrna&quot;</span>),</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">          <span class="dt">values=</span>gene_names,</a>
<a class="sourceLine" id="cb6-5" data-line-number="5">          <span class="dt">filters=</span><span class="st">&quot;refseq_mrna&quot;</span>,</a>
<a class="sourceLine" id="cb6-6" data-line-number="6">          <span class="dt">mart=</span>ensembl)</a>
<a class="sourceLine" id="cb6-7" data-line-number="7"></a>
<a class="sourceLine" id="cb6-8" data-line-number="8"><span class="co"># Create gene to GO id mapping</span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9">gene_id_go_mapping =<span class="st"> </span><span class="kw">create_go_term_mapping</span>(genes)</a></code></pre></div>
<p>Once the gene ID to GO mapping list is created, <code>moanin</code> provides a simple interface to <code>topGO</code> to fetch enriched GO terms.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">cluster =<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">ontologies =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;BP&quot;</span>, <span class="st">&quot;NF&quot;</span>, <span class="st">&quot;CC&quot;</span>)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="co"># It's taking too long to compile with everything. Just limit ourselves for</span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co"># now to BP</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5">ontologies =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;BP&quot;</span>)</a>
<a class="sourceLine" id="cb7-6" data-line-number="6">all_go_terms =<span class="st"> </span><span class="kw">data.frame</span>()</a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="cf">for</span>(cluster <span class="cf">in</span> clusters){</a>
<a class="sourceLine" id="cb7-8" data-line-number="8">    <span class="cf">for</span>(ontology <span class="cf">in</span> ontologies){ </a>
<a class="sourceLine" id="cb7-9" data-line-number="9">    scores =<span class="st"> </span><span class="kw">as.numeric</span>(labels <span class="op">==</span><span class="st"> </span>cluster)</a>
<a class="sourceLine" id="cb7-10" data-line-number="10">    <span class="kw">names</span>(scores) =<span class="st"> </span>gene_names</a>
<a class="sourceLine" id="cb7-11" data-line-number="11">    go_terms_enriched =<span class="st"> </span><span class="kw">find_enriched_go_terms</span>(</a>
<a class="sourceLine" id="cb7-12" data-line-number="12">        scores,</a>
<a class="sourceLine" id="cb7-13" data-line-number="13">        gene_id_go_mapping, <span class="dt">ontology=</span>ontology)</a>
<a class="sourceLine" id="cb7-14" data-line-number="14"></a>
<a class="sourceLine" id="cb7-15" data-line-number="15">    <span class="cf">if</span>((<span class="kw">dim</span>(all_go_terms)[<span class="dv">1</span>] <span class="op">==</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&amp;</span><span class="st"> </span>(<span class="kw">dim</span>(go_terms_enriched)[<span class="dv">1</span>] <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>)){</a>
<a class="sourceLine" id="cb7-16" data-line-number="16">        go_terms_enriched[<span class="st">&quot;Cluster&quot;</span>] =<span class="st"> </span>cluster</a>
<a class="sourceLine" id="cb7-17" data-line-number="17">        go_terms_enriched[<span class="st">&quot;Ontology&quot;</span>] =<span class="st"> </span>ontology</a>
<a class="sourceLine" id="cb7-18" data-line-number="18">        all_go_terms =<span class="st"> </span>go_terms_enriched</a>
<a class="sourceLine" id="cb7-19" data-line-number="19">    }<span class="cf">else</span> <span class="cf">if</span>(<span class="kw">dim</span>(go_terms_enriched)[<span class="dv">1</span>] <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>){</a>
<a class="sourceLine" id="cb7-20" data-line-number="20">        go_terms_enriched[<span class="st">&quot;Cluster&quot;</span>] =<span class="st"> </span>cluster</a>
<a class="sourceLine" id="cb7-21" data-line-number="21">        go_terms_enriched[<span class="st">&quot;Ontology&quot;</span>] =<span class="st"> </span>ontology</a>
<a class="sourceLine" id="cb7-22" data-line-number="22">        all_go_terms =<span class="st"> </span><span class="kw">merge</span>(all_go_terms, go_terms_enriched, <span class="dt">all=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb7-23" data-line-number="23">    }</a>
<a class="sourceLine" id="cb7-24" data-line-number="24">    }</a>
<a class="sourceLine" id="cb7-25" data-line-number="25">}</a></code></pre></div>
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
GO.ID
</th>
<th style="text-align:left;">
Term
</th>
<th style="text-align:right;">
Annotated
</th>
<th style="text-align:right;">
Significant
</th>
<th style="text-align:right;">
Expected
</th>
<th style="text-align:left;">
resultFisher
</th>
<th style="text-align:right;">
resultFisher_padj
</th>
<th style="text-align:right;">
Cluster
</th>
<th style="text-align:left;">
Ontology
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:left;">
<a href="GO:0007186" class="uri">GO:0007186</a>
</td>
<td style="text-align:left;">
G-protein coupled receptor signaling pat…
</td>
<td style="text-align:right;">
355
</td>
<td style="text-align:right;">
354
</td>
<td style="text-align:right;">
341.86
</td>
<td style="text-align:left;">
1.3e-05
</td>
<td style="text-align:right;">
0.0392
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
BP
</td>
</tr>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
<a href="GO:0007186" class="uri">GO:0007186</a>
</td>
<td style="text-align:left;">
G-protein coupled receptor signaling pat…
</td>
<td style="text-align:right;">
355
</td>
<td style="text-align:right;">
349
</td>
<td style="text-align:right;">
331.10
</td>
<td style="text-align:left;">
4.9e-06
</td>
<td style="text-align:right;">
0.0074
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
BP
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
<a href="GO:0007608" class="uri">GO:0007608</a>
</td>
<td style="text-align:left;">
sensory perception of smell
</td>
<td style="text-align:right;">
174
</td>
<td style="text-align:right;">
174
</td>
<td style="text-align:right;">
162.28
</td>
<td style="text-align:left;">
4.1e-06
</td>
<td style="text-align:right;">
0.0074
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
BP
</td>
</tr>
</tbody>
</table>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
