<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Downstream analysis: A pipeline to analyse time-course gene expression data</title>

<script>$(document).ready(function(){
    if (typeof $('[data-toggle="tooltip"]').tooltip === 'function') {
        $('[data-toggle="tooltip"]').tooltip();
    }
    if ($('[data-toggle="popover"]').popover === 'function') {
        $('[data-toggle="popover"]').popover();
    }
});
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Downstream analysis: A pipeline to analyse time-course gene expression data</h1>



<pre><code>## Warning: no function found corresponding to methods exports from 'XVector'
## for: 'concatenateObjects'</code></pre>
<div id="downstream-analysis-of-clusters." class="section level2">
<h2>Downstream analysis of clusters.</h2>
<p>Once good clusters are obtained, the next step is to leverage the clustering to ease interpretation. Classic enrichment analysis step can then be performed in the gene set defined in each cluster: KEGG pathway enrichment analysis, GO term enrichment analysis, motif enrichment analysis, etc.</p>
<p>First, let us clean up the gene obtained and only select the genes we are going to use in the enrichment analysis. One can either use the whole set of genes, only the set of differentially expressed genes in each cluster, or a subset of genes that fit well to a cluster (based on some criterion).</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">gene_names =<span class="st"> </span><span class="kw">row.names</span>(labels)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">gene_names =<span class="st"> </span>gene_names[<span class="op">!</span><span class="kw">is.na</span>(labels)]</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">labels =<span class="st"> </span><span class="kw">as.vector</span>(labels[<span class="op">!</span><span class="kw">is.na</span>(labels)])</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="kw">names</span>(labels) =<span class="st"> </span>gene_names</a></code></pre></div>
<div id="finding-enriched-pathways-using-biomart-and-keggprofile" class="section level3">
<h3>Finding enriched pathways using <code>biomaRt</code> and <code>KEGGprofile</code></h3>
<p>Let us first tackle the case of pathway enrichment analysis. We will leverage the packages <code>biomaRt</code> and <code>KEGGprofile</code> for this step. <code>KEGGprofile</code> is a package that easily allows to perform pathway enrichment analysis on a set of genes labeled with the ensembl annotation. We thus need to convert the gene names into the appropriate format. This is where <code>biomaRt</code> comes in handy: it enables easy conversion from one gene annotation to another. Here, we will use it to convert the gene names from the Refseq annotation to the ensembl one.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">library</span>(biomaRt)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">ensembl =<span class="st"> </span>biomaRt<span class="op">::</span><span class="kw">useMart</span>(<span class="st">&quot;ensembl&quot;</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">ensembl =<span class="st"> </span>biomaRt<span class="op">::</span><span class="kw">useDataset</span>(<span class="st">&quot;mmusculus_gene_ensembl&quot;</span>, <span class="dt">mart=</span>ensembl)</a></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">cluster =<span class="st"> </span><span class="dv">4</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">genes =<span class="st"> </span>gene_names[labels <span class="op">==</span><span class="st"> </span>cluster]</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co"># convert gene names</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5">genes =<span class="st"> </span>biomaRt<span class="op">::</span><span class="kw">getBM</span>(<span class="dt">attributes=</span><span class="kw">c</span>(<span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;entrezgene&quot;</span>),</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">        <span class="dt">filters=</span><span class="st">&quot;refseq_mrna&quot;</span>, <span class="dt">values=</span>genes,</a>
<a class="sourceLine" id="cb4-7" data-line-number="7">        <span class="dt">mart=</span>ensembl)[<span class="st">&quot;entrezgene&quot;</span>]</a>
<a class="sourceLine" id="cb4-8" data-line-number="8">genes =<span class="st"> </span><span class="kw">as.vector</span>(<span class="kw">unlist</span>(genes))</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">pathways =<span class="st"> </span>KEGGprofile<span class="op">::</span><span class="kw">find_enriched_pathway</span>(</a>
<a class="sourceLine" id="cb4-10" data-line-number="10">    genes, <span class="dt">species=</span><span class="st">&quot;mmu&quot;</span>,</a>
<a class="sourceLine" id="cb4-11" data-line-number="11">    <span class="dt">download_latest=</span><span class="ot">TRUE</span>)<span class="op">$</span>stastic</a></code></pre></div>
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Pathway
</th>
<th style="text-align:right;">
Percentage
</th>
<th style="text-align:right;">
Adj. p-value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
00010
</td>
<td style="text-align:left;">
Glycolysis / Gluconeogenesis
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
0.0282
</td>
</tr>
<tr>
<td style="text-align:left;">
01100
</td>
<td style="text-align:left;">
Metabolic pathways
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0.0282
</td>
</tr>
<tr>
<td style="text-align:left;">
01200
</td>
<td style="text-align:left;">
Carbon metabolism
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
0.0282
</td>
</tr>
<tr>
<td style="text-align:left;">
04540
</td>
<td style="text-align:left;">
Gap junction
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
0.0282
</td>
</tr>
<tr>
<td style="text-align:left;">
04910
</td>
<td style="text-align:left;">
Insulin signaling pathway
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0.0282
</td>
</tr>
<tr>
<td style="text-align:left;">
05168
</td>
<td style="text-align:left;">
Herpes simplex virus 1 infection
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.0282
</td>
</tr>
<tr>
<td style="text-align:left;">
05225
</td>
<td style="text-align:left;">
Hepatocellular carcinoma
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0.0396
</td>
</tr>
<tr>
<td style="text-align:left;">
04066
</td>
<td style="text-align:left;">
HIF-1 signaling pathway
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0.0431
</td>
</tr>
</tbody>
</table>
</div>
<div id="finding-enriched-go-terms" class="section level3">
<h3>Finding enriched GO terms</h3>
<p>To find GO terms, we use <code>biomaRt</code> to find the mapping between GO terms and gene mapping. The GO enrichment library <code>topGO</code> expects the GO term to gene mapping to be a list where each item is a mapping between a gene name and a GO term ID vector.</p>
<pre class="text"><code>$NM_199153
[1] &quot;GO:0016020&quot; &quot;GO:0016021&quot; &quot;GO:0007186&quot; &quot;GO:0004930&quot; &quot;GO:0007165&quot;
[6] &quot;GO:0050896&quot; &quot;GO:0050909&quot;

$NM_201361
 [1] &quot;GO:0016020&quot; &quot;GO:0016021&quot; &quot;GO:0003674&quot; &quot;GO:0008150&quot; &quot;GO:0005794&quot;
 [6] &quot;GO:0005829&quot; &quot;GO:0005737&quot; &quot;GO:0005856&quot; &quot;GO:0005874&quot; &quot;GO:0005739&quot;
[11] &quot;GO:0005819&quot; &quot;GO:0000922&quot; &quot;GO:0072686&quot;</code></pre>
<p><code>biomaRt</code> queries results a matrix with two named columns of gene names and GO term ID.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">library</span>(topGO)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">genes =<span class="st"> </span>biomaRt<span class="op">::</span><span class="kw">getBM</span>(<span class="dt">attributes=</span><span class="kw">c</span>(<span class="st">&quot;go_id&quot;</span>, <span class="st">&quot;refseq_mrna&quot;</span>),</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">          <span class="dt">values=</span>gene_names,</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">          <span class="dt">filters=</span><span class="st">&quot;refseq_mrna&quot;</span>,</a>
<a class="sourceLine" id="cb6-5" data-line-number="5">          <span class="dt">mart=</span>ensembl)</a>
<a class="sourceLine" id="cb6-6" data-line-number="6"></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="co"># Create gene to GO id mapping</span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8">gene_id_go_mapping =<span class="st"> </span><span class="kw">create_go_term_mapping</span>(genes)</a></code></pre></div>
<p>Once the gene ID to GO mapping list is created, <code>moanin</code> provides a simple interface to <code>topGO</code> to fetch enriched GO terms. Here, we show an example of running a GO term enrichment on the “Biological process” ontology (BP).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">assignments =<span class="st"> </span>labels <span class="op">==</span><span class="st"> </span>cluster</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"></a>
<a class="sourceLine" id="cb7-3" data-line-number="3">go_terms_enriched =<span class="st"> </span><span class="kw">find_enriched_go_terms</span>(</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">    assignments,</a>
<a class="sourceLine" id="cb7-5" data-line-number="5">    gene_id_go_mapping, <span class="dt">ontology=</span><span class="st">&quot;BP&quot;</span>)</a></code></pre></div>
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
GO.ID
</th>
<th style="text-align:left;">
Term
</th>
<th style="text-align:right;">
Annotated
</th>
<th style="text-align:right;">
Significant
</th>
<th style="text-align:right;">
Expected
</th>
<th style="text-align:left;">
resultFisher
</th>
<th style="text-align:right;">
resultFisher_padj
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
<a href="GO:0007610" class="uri">GO:0007610</a>
</td>
<td style="text-align:left;">
behavior
</td>
<td style="text-align:right;">
299
</td>
<td style="text-align:right;">
299
</td>
<td style="text-align:right;">
289.24
</td>
<td style="text-align:left;">
0.00013
</td>
<td style="text-align:right;">
0.4278
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0072503" class="uri">GO:0072503</a>
</td>
<td style="text-align:left;">
cellular divalent inorganic cation homeo…
</td>
<td style="text-align:right;">
258
</td>
<td style="text-align:right;">
258
</td>
<td style="text-align:right;">
249.58
</td>
<td style="text-align:left;">
0.00017
</td>
<td style="text-align:right;">
0.4278
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0019932" class="uri">GO:0019932</a>
</td>
<td style="text-align:left;">
second-messenger-mediated signaling
</td>
<td style="text-align:right;">
222
</td>
<td style="text-align:right;">
221
</td>
<td style="text-align:right;">
214.75
</td>
<td style="text-align:left;">
0.00074
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0007186" class="uri">GO:0007186</a>
</td>
<td style="text-align:left;">
G protein-coupled receptor signaling pat…
</td>
<td style="text-align:right;">
515
</td>
<td style="text-align:right;">
509
</td>
<td style="text-align:right;">
498.19
</td>
<td style="text-align:left;">
0.00166
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0072511" class="uri">GO:0072511</a>
</td>
<td style="text-align:left;">
divalent inorganic cation transport
</td>
<td style="text-align:right;">
262
</td>
<td style="text-align:right;">
260
</td>
<td style="text-align:right;">
253.45
</td>
<td style="text-align:left;">
0.00183
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0032103" class="uri">GO:0032103</a>
</td>
<td style="text-align:left;">
positive regulation of response to exter…
</td>
<td style="text-align:right;">
187
</td>
<td style="text-align:right;">
187
</td>
<td style="text-align:right;">
180.90
</td>
<td style="text-align:left;">
0.00189
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0030031" class="uri">GO:0030031</a>
</td>
<td style="text-align:left;">
cell projection assembly
</td>
<td style="text-align:right;">
302
</td>
<td style="text-align:right;">
300
</td>
<td style="text-align:right;">
292.14
</td>
<td style="text-align:left;">
0.00250
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:1904062" class="uri">GO:1904062</a>
</td>
<td style="text-align:left;">
regulation of cation transmembrane trans…
</td>
<td style="text-align:right;">
175
</td>
<td style="text-align:right;">
175
</td>
<td style="text-align:right;">
169.29
</td>
<td style="text-align:left;">
0.00284
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0010959" class="uri">GO:0010959</a>
</td>
<td style="text-align:left;">
regulation of metal ion transport
</td>
<td style="text-align:right;">
232
</td>
<td style="text-align:right;">
231
</td>
<td style="text-align:right;">
224.43
</td>
<td style="text-align:left;">
0.00370
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0043207" class="uri">GO:0043207</a>
</td>
<td style="text-align:left;">
response to external biotic stimulus
</td>
<td style="text-align:right;">
536
</td>
<td style="text-align:right;">
528
</td>
<td style="text-align:right;">
518.51
</td>
<td style="text-align:left;">
0.00448
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0010469" class="uri">GO:0010469</a>
</td>
<td style="text-align:left;">
regulation of signaling receptor activit…
</td>
<td style="text-align:right;">
268
</td>
<td style="text-align:right;">
266
</td>
<td style="text-align:right;">
259.25
</td>
<td style="text-align:left;">
0.00641
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0051091" class="uri">GO:0051091</a>
</td>
<td style="text-align:left;">
positive regulation of DNA-binding trans…
</td>
<td style="text-align:right;">
149
</td>
<td style="text-align:right;">
149
</td>
<td style="text-align:right;">
144.14
</td>
<td style="text-align:left;">
0.00683
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0051146" class="uri">GO:0051146</a>
</td>
<td style="text-align:left;">
striated muscle cell differentiation
</td>
<td style="text-align:right;">
146
</td>
<td style="text-align:right;">
146
</td>
<td style="text-align:right;">
141.23
</td>
<td style="text-align:left;">
0.00756
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0097305" class="uri">GO:0097305</a>
</td>
<td style="text-align:left;">
response to alcohol
</td>
<td style="text-align:right;">
144
</td>
<td style="text-align:right;">
144
</td>
<td style="text-align:right;">
139.30
</td>
<td style="text-align:left;">
0.00809
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0032409" class="uri">GO:0032409</a>
</td>
<td style="text-align:left;">
regulation of transporter activity
</td>
<td style="text-align:right;">
135
</td>
<td style="text-align:right;">
135
</td>
<td style="text-align:right;">
130.59
</td>
<td style="text-align:left;">
0.01096
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0032970" class="uri">GO:0032970</a>
</td>
<td style="text-align:left;">
regulation of actin filament-based proce…
</td>
<td style="text-align:right;">
215
</td>
<td style="text-align:right;">
214
</td>
<td style="text-align:right;">
207.98
</td>
<td style="text-align:left;">
0.01104
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0051259" class="uri">GO:0051259</a>
</td>
<td style="text-align:left;">
protein complex oligomerization
</td>
<td style="text-align:right;">
333
</td>
<td style="text-align:right;">
329
</td>
<td style="text-align:right;">
322.13
</td>
<td style="text-align:left;">
0.01396
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0042060" class="uri">GO:0042060</a>
</td>
<td style="text-align:left;">
wound healing
</td>
<td style="text-align:right;">
220
</td>
<td style="text-align:right;">
218
</td>
<td style="text-align:right;">
212.82
</td>
<td style="text-align:left;">
0.01621
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0043254" class="uri">GO:0043254</a>
</td>
<td style="text-align:left;">
regulation of protein complex assembly
</td>
<td style="text-align:right;">
245
</td>
<td style="text-align:right;">
243
</td>
<td style="text-align:right;">
237.00
</td>
<td style="text-align:left;">
0.01627
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0050920" class="uri">GO:0050920</a>
</td>
<td style="text-align:left;">
regulation of chemotaxis
</td>
<td style="text-align:right;">
131
</td>
<td style="text-align:right;">
131
</td>
<td style="text-align:right;">
126.72
</td>
<td style="text-align:left;">
0.01747
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0007283" class="uri">GO:0007283</a>
</td>
<td style="text-align:left;">
spermatogenesis
</td>
<td style="text-align:right;">
228
</td>
<td style="text-align:right;">
226
</td>
<td style="text-align:right;">
220.56
</td>
<td style="text-align:left;">
0.01870
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0002685" class="uri">GO:0002685</a>
</td>
<td style="text-align:left;">
regulation of leukocyte migration
</td>
<td style="text-align:right;">
116
</td>
<td style="text-align:right;">
116
</td>
<td style="text-align:right;">
112.21
</td>
<td style="text-align:left;">
0.02077
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0051051" class="uri">GO:0051051</a>
</td>
<td style="text-align:left;">
negative regulation of transport
</td>
<td style="text-align:right;">
275
</td>
<td style="text-align:right;">
272
</td>
<td style="text-align:right;">
266.02
</td>
<td style="text-align:left;">
0.02210
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0046395" class="uri">GO:0046395</a>
</td>
<td style="text-align:left;">
carboxylic acid catabolic process
</td>
<td style="text-align:right;">
114
</td>
<td style="text-align:right;">
114
</td>
<td style="text-align:right;">
110.28
</td>
<td style="text-align:left;">
0.02221
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0043901" class="uri">GO:0043901</a>
</td>
<td style="text-align:left;">
negative regulation of multi-organism pr…
</td>
<td style="text-align:right;">
107
</td>
<td style="text-align:right;">
107
</td>
<td style="text-align:right;">
103.51
</td>
<td style="text-align:left;">
0.02810
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0030308" class="uri">GO:0030308</a>
</td>
<td style="text-align:left;">
negative regulation of cell growth
</td>
<td style="text-align:right;">
107
</td>
<td style="text-align:right;">
107
</td>
<td style="text-align:right;">
103.51
</td>
<td style="text-align:left;">
0.02810
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0051241" class="uri">GO:0051241</a>
</td>
<td style="text-align:left;">
negative regulation of multicellular org…
</td>
<td style="text-align:right;">
673
</td>
<td style="text-align:right;">
660
</td>
<td style="text-align:right;">
651.03
</td>
<td style="text-align:left;">
0.02865
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0043062" class="uri">GO:0043062</a>
</td>
<td style="text-align:left;">
extracellular structure organization
</td>
<td style="text-align:right;">
161
</td>
<td style="text-align:right;">
160
</td>
<td style="text-align:right;">
155.75
</td>
<td style="text-align:left;">
0.02977
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0097193" class="uri">GO:0097193</a>
</td>
<td style="text-align:left;">
intrinsic apoptotic signaling pathway
</td>
<td style="text-align:right;">
161
</td>
<td style="text-align:right;">
160
</td>
<td style="text-align:right;">
155.75
</td>
<td style="text-align:left;">
0.02977
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0042493" class="uri">GO:0042493</a>
</td>
<td style="text-align:left;">
response to drug
</td>
<td style="text-align:right;">
576
</td>
<td style="text-align:right;">
565
</td>
<td style="text-align:right;">
557.20
</td>
<td style="text-align:left;">
0.02977
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0006954" class="uri">GO:0006954</a>
</td>
<td style="text-align:left;">
inflammatory response
</td>
<td style="text-align:right;">
372
</td>
<td style="text-align:right;">
367
</td>
<td style="text-align:right;">
359.86
</td>
<td style="text-align:left;">
0.03079
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0043068" class="uri">GO:0043068</a>
</td>
<td style="text-align:left;">
positive regulation of programmed cell d…
</td>
<td style="text-align:right;">
355
</td>
<td style="text-align:right;">
350
</td>
<td style="text-align:right;">
343.41
</td>
<td style="text-align:left;">
0.03380
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0002274" class="uri">GO:0002274</a>
</td>
<td style="text-align:left;">
myeloid leukocyte activation
</td>
<td style="text-align:right;">
116
</td>
<td style="text-align:right;">
116
</td>
<td style="text-align:right;">
112.21
</td>
<td style="text-align:left;">
0.03533
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0046677" class="uri">GO:0046677</a>
</td>
<td style="text-align:left;">
response to antibiotic
</td>
<td style="text-align:right;">
207
</td>
<td style="text-align:right;">
205
</td>
<td style="text-align:right;">
200.24
</td>
<td style="text-align:left;">
0.03549
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0002699" class="uri">GO:0002699</a>
</td>
<td style="text-align:left;">
positive regulation of immune effector p…
</td>
<td style="text-align:right;">
137
</td>
<td style="text-align:right;">
136
</td>
<td style="text-align:right;">
132.53
</td>
<td style="text-align:left;">
0.03789
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0044782" class="uri">GO:0044782</a>
</td>
<td style="text-align:left;">
cilium organization
</td>
<td style="text-align:right;">
197
</td>
<td style="text-align:right;">
195
</td>
<td style="text-align:right;">
190.57
</td>
<td style="text-align:left;">
0.04136
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0002684" class="uri">GO:0002684</a>
</td>
<td style="text-align:left;">
positive regulation of immune system pro…
</td>
<td style="text-align:right;">
495
</td>
<td style="text-align:right;">
489
</td>
<td style="text-align:right;">
478.84
</td>
<td style="text-align:left;">
0.04452
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0002262" class="uri">GO:0002262</a>
</td>
<td style="text-align:left;">
myeloid cell homeostasis
</td>
<td style="text-align:right;">
92
</td>
<td style="text-align:right;">
92
</td>
<td style="text-align:right;">
89.00
</td>
<td style="text-align:left;">
0.04649
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0007163" class="uri">GO:0007163</a>
</td>
<td style="text-align:left;">
establishment or maintenance of cell pol…
</td>
<td style="text-align:right;">
111
</td>
<td style="text-align:right;">
110
</td>
<td style="text-align:right;">
107.38
</td>
<td style="text-align:left;">
0.04668
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0040013" class="uri">GO:0040013</a>
</td>
<td style="text-align:left;">
negative regulation of locomotion
</td>
<td style="text-align:right;">
190
</td>
<td style="text-align:right;">
188
</td>
<td style="text-align:right;">
183.80
</td>
<td style="text-align:left;">
0.04923
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="GO:0061448" class="uri">GO:0061448</a>
</td>
<td style="text-align:left;">
connective tissue development
</td>
<td style="text-align:right;">
143
</td>
<td style="text-align:right;">
142
</td>
<td style="text-align:right;">
138.33
</td>
<td style="text-align:left;">
0.04940
</td>
<td style="text-align:right;">
1.0000
</td>
</tr>
</tbody>
</table>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
