\documentclass[11 pts]{article}

\usepackage[utf8]{inputenc}
\usepackage{multirow}
\usepackage{graphicx}
\usepackage{booktabs}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{commath}

\usepackage[font={small}]{caption}
\usepackage{longtable}

\newcommand{\todo}[1]{\textbf{[TODO\@: #1]}}
\newcommand{\note}[1]{\textbf{[NOTE\@: #1]}}
\newcommand{\fixme}[1]{\textbf{[FIXME\@: #1]}}


\begin{document}
\title{A pipeline to analyse transcriptomic time-course data}
\maketitle


\begin{abstract}

The phenotypic diversity of cells is governed by a complex equilibrium between
their genetic identity and their environmental interactions, and understanding
the dynamics of gene expression is a fundamental question of biology.Â However,
analysing time-course transcriptomic data raises challenging statistical and
computational questions, requiring the development of novel methods and
software. We present a case study of the lifetime of S. bicolor under
droughted and watered condition using an integrated workflow providing a
step-by-step tutorial to the methodology and associated software for the
following four main tasks.

Novel single-cell transcriptome sequencing assays allow researchers to measure
gene expression levels at the resolution of single cells and offer the
unprecendented opportunity to investigate at the molecular level fundamental
biological questions, such as stem cell differentiation or the discovery and
characterization of rare cell types. However, such assays raise challenging
statistical and computational questions and require the development of novel
methodology and software. Using stem cell differentiation in the mouse
olfactory epithelium as a case study, this integrated workflow provides a
step-by-step tutorial to the methodology and associated software for the
following four main tasks: (1) dimensionality reduction accounting for zero
inflation and over dispersion and adjusting for gene and cell-level
covariates; (2) cell clustering using resampling-based sequential ensemble
clustering; (3) inference of cell lineages and pseudotimes; and (4)
differential expression analysis along lineages
\end{abstract}


\section{Introduction}

\begin{figure}
\caption{Workflow for analysing time-course RNASeq experiments}
\end{figure}

\section{Results}
\subsection{Quality control analysis}

\begin{table}
\caption{Summary of time-course RNASeq experiments and results of DE Analysis}
\end{table}


\subsection{Normalization}

\subsection{Differential expression analysis}

\subsection{Clustering}

\subsection{Downstream analysis}


\section{Conclusion}


\end{document}
