\documentclass[11 pts]{article}

\usepackage[utf8]{inputenc}
\usepackage{multirow}
\usepackage{graphicx}
\usepackage{booktabs}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{commath}

\usepackage[font={small}]{caption}
\usepackage{longtable}
\usepackage{color}
\newcommand{\todo}[1]{\textbf{[TODO\@: #1]}}
\newcommand{\note}[1]{\textbf{[NOTE\@: #1]}}
\newcommand{\fixme}[1]{\textbf{[FIXME\@: #1]}}
\newcommand{\ep}[1]{\textcolor{cyan}{[EP\@: #1]}}
\newcommand{\xxx}[1]{\textcolor{red}{XXX}}


\begin{document}
\title{A pipeline to analyse transcriptomic time-course data}
\maketitle


\begin{abstract}

The phenotypic diversity of cells is governed by a complex equilibrium between
their genetic identity and their environmental interactions, and understanding
the dynamics of gene expression is a fundamental question of biology.Â However,
analysing time-course transcriptomic data raises challenging statistical and
computational questions, requiring the development of novel methods and
software. We present a case study of the lifetime of S. bicolor under
droughted and watered condition using an integrated workflow providing a
step-by-step tutorial to the methodology and associated software for the
following four main tasks.

Novel single-cell transcriptome sequencing assays allow researchers to measure
gene expression levels at the resolution of single cells and offer the
unprecendented opportunity to investigate at the molecular level fundamental
biological questions, such as stem cell differentiation or the discovery and
characterization of rare cell types. However, such assays raise challenging
statistical and computational questions and require the development of novel
methodology and software. Using stem cell differentiation in the mouse
olfactory epithelium as a case study, this integrated workflow provides a
step-by-step tutorial to the methodology and associated software for the
following four main tasks: (1) dimensionality reduction accounting for zero
inflation and over dispersion and adjusting for gene and cell-level
covariates; (2) cell clustering using resampling-based sequential ensemble
clustering; (3) inference of cell lineages and pseudotimes; and (4)
differential expression analysis along lineages

\ep{I assume this is, at this point, two different possible abstracts?}
\end{abstract}


\section{Introduction}

Gene expression studies provide simultaneous quantification of the level of mRNA from all genes in a sample. High-throughput studies of gene expression have a long history, starting with microarray technologies in the 1990s through to single-cell technologies. While many expression studies are designed to compare the gene expression in distinct groups, there is also a long history of time-course expression studies, where the the gene expression is compared across time by measuring mRNA levels from different samples across time.\footnote{Because the collection of the mRNA is often destructive, samples at different time points are generally from different biological samples; longitudinal studies, for example tracking the same subject over time, are certainly possible in certain settings, but not directly considered here.} Such time course studies can vary from measuring a few distinct time points, to sampling ten to twenty time points. Many longer time series are particularly interested in investigating development over time. More recently, single-cell studies track single cells through their development, and a single cell is measured at a particular moment in its developmental progression -- a value that is not know but estimated from the data as its ``pseudo-time''.

While there are many methods that have been proposed for discrete aspects of time course data, the entire workflow for analysis of such data remains difficult, particularly for long, developmental time series. Most methods proposed for time course data are concerned with detecting genes that are changing over time (differential expression analysis), examples being \texttt{edge}\todo[Other methods to mention: impulse models, what else? Citations!]. However, with long time course datasets, particularly in developmental systems, a massive number of genes will show \emph{some} change. For example, in the \xxx system, over \xxx\% of genes are shown to be changing over time. The task in these settings is often not to detect changes in genes, but to categorize into biologically interpretable patterns the vast number of changes discovered. 

We present here a workflow for such an analysis that consists of \xxx main parts (Figure \ref{fig:workflow}):
\begin{enumerate}
	\item Identification of genes that are differentially expressed
	\item Clustering of genes into distinct temporal patterns
	\item Biological interpretation of the clusters
\end{enumerate}
\ep{This is probably not best enumeration of steps, but need something...}

\begin{figure}
\caption{Workflow for analysing time-course RNASeq experiments}
\label{fig:workflow}
\end{figure}

This workflow represents an integration of both novel implementations of previously established methods and new methodologies for the settings of developmental time series. We provide the various steps of the workflow as functions in a R package, \citettt{\xxx}, along with command-line tools for running the entire workflow in a single command. 

\section{Results}
\subsection{Quality control analysis}

\begin{table}
\caption{Summary of time-course RNASeq experiments and results of DE Analysis}
\end{table}


\subsection{Normalization}

\subsection{Differential expression analysis}

Several techniques for mRNA-expression analysis were

\subsection{Clustering}

\subsection{Downstream analysis}


\section{Conclusion}


\end{document}
