---
title: "Introduction: a pipeline to analyse time-course gene expression data"
output: rmarkdown::html_vignette
---

# Abstract


The phenotypic diversity of cells is governed by a complex equilibrium between
their genetic identity and their environmental interactions: Understanding the
dynamics of gene expression is a fundamental question of biology.Â However,
analysing time-course transcriptomic data raises unique challenging
statistical and computational questions, requiring the development of novel
methods and software.  Using as case study time-course transcriptomics data
from mice exposed to different strains of influenza, ihis workflows provides a
step-by-step tutorial of the methodology used to analyse time-course data:
(1) normalization of the micro-array dataset; (2) differential expression
analysis using functional data analysis; (3) clustering fo time-course data;
(4) interpreting clusters with GO term and KEGG pathway enrichment analysis.


# Introduction

Time-course gene expression analysis workflows include steps used more broadly
for gene expression data analysis such as quality control (QC), exploratory
data analysis (EDA), pre-processing, normalization, and visualization. Yet, a
typical gene expression data analysis workflow is not entirely adequate for
analysing time-course data. In particular, differential expression analysis
needs to be adapted to time-course data, and often needs to be followed by a
clustering step to group genes sharing similar patterns of expression.

Here, we propose an integrated workflow for downstream analysis of time-course
data, with the following main steps: (1) quality control and normalization;
(2) differential expression analysis of time-course data; (3) clustering; (4)
downstream analysis of clusters with GO term, pathway, and motif enrichment.

This workflow relies on several standard packages for analysing gene
expression data, some specific for time-course data, others broadly used by
the community. We also provide alongside this workflow a package that nicely
wraps all calls to dependencies with a common API by defining a model object
holding all necessary metadata to perform all steps of the pipeline.

# Analysis of the dynamical response of mouse lung tissue to influenza

## Overview

This workflow is illustrated using data from a micro-array time-course
experiment, exposing mice to three different strains of influenza, and
collecting lung tissue during 14 time-points after infection (0, 3, 6, 9, 12,
18, 24, 30, 36, 48, 60 hours, then 3, 5, and 7 days later). The three strains
of influenza used in the study are (1) a low pathogenicity seasonal H1N1
influenza virus (A/Kawasaki/UTK4/2009 [H1N1]), a mildly pathogenic virus from
the 2009 pandemic season (A/California/04/2009 [H1N1]), and a highly
pathogenic H5N1 avian influenza virus (A/Vietnam/1203/2004 [H5N1]. Mice were
injected with 10^5 PFU of each virus. An adidtional 42 mice were injected with
a lower dose of the Vietnam avian influenza virus (10^3 PFU).

The authors find that 

**XXX Add summary of the key findings of the the paper**

