---
title: "Introduction: a pipeline to analyse time-course gene expression data"
output: rmarkdown::html_vignette
---

# Abstract


The phenotypic diversity of cells is governed by a complex equilibrium between
their genetic identity and their environmental interactions: Understanding the
dynamics of gene expression is a fundamental question of biology.Â However,
analysing time-course transcriptomic data raises unique challenging
statistical and computational questions, requiring the development of novel
methods and software.  Using as case study time-course transcriptomics data
from mice exposed to different strains of influenza, ihis workflows provides a
step-by-step tutorial of the methodology used to analyse time-course data:
(1) normalization of the micro-array dataset; (2) differential expression
analysis using functional data analysis; (3) clustering fo time-course data;
(4) interpreting clusters with GO term and KEGG pathway enrichment analysis.


# Introduction

Gene expression studies provide simultaneous quantification of the level of
mRNA from all genes in a sample. High-throughput studies of gene expression
have a long history, starting with microarray technologies in the 1990s
through to single-cell technologies. While many expression studies are
designed to compare the gene expression in distinct groups, there is also a
long history of time-course expression studies, where the the gene expression
is compared across time by measuring mRNA levels from different samples across
time.\footnote{Because the collection of the mRNA is often destructive,
samples at different time points are generally from different biological
samples; longitudinal studies, for example tracking the same subject over
time, are certainly possible in certain settings, but not directly considered
here.} Such time course studies can vary from measuring a few distinct time
points, to sampling ten to twenty time points. Many longer time series are
particularly interested in investigating development over time. More recently,
single-cell studies track single cells through their development, and a single
cell is measured at a particular moment in its developmental progression -- a
value that is not know but estimated from the data as its "pseudo-time."

While there are many methods that have been proposed for discrete aspects of
time course data, the entire workflow for analysis of such data remains
difficult, particularly for long, developmental time series. Most methods
proposed for time course data are concerned with detecting genes that are
changing over time (differential expression analysis), examples being
\texttt{edge}\todo[Other methods to mention: impulse models, what else?
Citations!]. However, with long time course datasets, particularly in
developmental systems, a massive number of genes will show *some* change.
For example, in the \xxx system, over \xxx\% of genes are shown to be changing
over time. The task in these settings is often not to detect changes in genes,
but to categorize into biologically interpretable patterns the vast number of
changes discovered. 

We present here a workflow for such an analysis that consists of \xxx main
parts (Figure \ref{fig:workflow}):

- Identification of genes that are differentially expressed
- Clustering of genes into distinct temporal patterns
- Biological interpretation of the clusters

**FIXME EP This is probably not best enumeration of steps, but need
something...**

This workflow represents an integration of both novel implementations of
previously established methods and new methodologies for the settings of
developmental time series. We provide the various steps of the workflow as
functions in a R package, \citettt{\xxx}, along with command-line tools for
running the entire workflow in a single command. 


---
# Old intro

Time-course gene expression analysis workflows include steps used more broadly
for gene expression data analysis such as quality control (QC), exploratory
data analysis (EDA), pre-processing, normalization, and visualization. Yet, a
typical gene expression data analysis workflow is not entirely adequate for
analysing time-course data. In particular, differential expression analysis
needs to be adapted to time-course data, and often needs to be followed by a
clustering step to group genes sharing similar patterns of expression.

Here, we propose an integrated workflow for downstream analysis of time-course
data, with the following main steps: (1) quality control and normalization;
(2) differential expression analysis of time-course data; (3) clustering; (4)
downstream analysis of clusters with GO term, pathway, and motif enrichment.

This workflow relies on several standard packages for analysing gene
expression data, some specific for time-course data, others broadly used by
the community. We also provide alongside this workflow a package that nicely
wraps all calls to dependencies with a common API by defining a model object
holding all necessary metadata to perform all steps of the pipeline.
---

# Analysis of the dynamical response of mouse lung tissue to influenza

## Overview

This workflow is illustrated using data from a micro-array time-course
experiment, exposing mice to three different strains of influenza, and
collecting lung tissue during 14 time-points after infection (0, 3, 6, 9, 12,
18, 24, 30, 36, 48, 60 hours, then 3, 5, and 7 days later). The three strains
of influenza used in the study are (1) a low pathogenicity seasonal H1N1
influenza virus (A/Kawasaki/UTK4/2009 [H1N1]), a mildly pathogenic virus from
the 2009 pandemic season (A/California/04/2009 [H1N1]), and a highly
pathogenic H5N1 avian influenza virus (A/Vietnam/1203/2004 [H5N1]. Mice were
injected with 10^5 PFU of each virus. An adidtional 42 mice were injected with
a lower dose of the Vietnam avian influenza virus (10^3 PFU).

The authors find that 

**XXX Add summary of the key findings of the the paper**

