---
title: "Differential expression analysis: A pipeline to analyse time-course gene expression data"
output: rmarkdown::html_vignette
bibliography: bibliography.bib
---


```{r echo=FALSE, results="hide"}
# Reloading necessary data and preliminary results
library(devtools)
library(limma)
library(splines)
library(stats)
library(moanin)
library(knitr)
library(kableExtra)
source("utils.R")
```

```{r echo=FALSE}
# Set options
df = 6

# Preprocessing & filtering
is_count = FALSE
take_log = FALSE    
filter_expression = FALSE

# Differential expression analysis
timecourse_contrasts = c("M-K", "M-C", "M-VL", "M-VH")

# Clustering
filter_genes = TRUE
n_genes_to_keep = 5000
n_clusters = 20
percentage_genes_to_label = 0.5
```


```{r echo=FALSE, results="hide"}
# Set data files and options
data("shoemaker2015")
data = shoemaker2015$data
meta = shoemaker2015$meta
meta = droplevels(meta)

# Right now, the package doesn't contain the normalized data
```

```{r, echo=FALSE, results="hide"}
# XXX FIXME
data = read.table(".results/quantile_normalized.txt") 
source("utils.R")
```


## Differential expression analysis of time-course data

The next step is typically to run a differential expression analysis. Most of
`moanin`'s function rely on an object containing a number of information such
as the metadata and the splines model used.

```{r}
splines_model = moanin::create_splines_model(meta, degrees_of_freedom=6)
```


### Weekly differential expression analysis

`moanin` provides a simple interface to perform a time by time differential
expression analysis. Under the hood, it simply calls `limma` on the set of
contrasts provided. Here, we show an example between the control mouse and the
mouse infected with the high dose of the influenza strain A/Vietnam/1203/04
(H5N1), but the function works with any form contrasts.

```{r weekly_differential_analysis}
meta$WeeklyGroup = make.names(meta$Group:as.factor(meta$Timepoint))

# Define contrast	
contrasts = c("M.0-VL.0",
	      "M.3-VL.3",
	      "M.6-VL.6",
	      "M.9-VL.9",
	      "M.12-VL.12",
	      "M.18-VL.18",
	      "M.24-VL.24",
	      "M.30-VL.30",
	      "M.36-VL.36",
	      "M.48-VL.48",
	      "M.60-VL.60",
	      "M.72-VL.72",
	      "M.120-VL.120",
	      "M.168-VL.168")
weekly_de_analysis =  moanin::weekly_differential_expression_analysis(
     data, splines_model, contrasts,
     use_voom_weights=FALSE)
```

```{r echo=FALSE, results="hide"}
# Hide those results
# Run the DE analysis on the rest of the influenza strain.
contrasts = c("M.0-K.0",
	      "M.3-K.3",
	      "M.6-K.6",
	      "M.9-K.9",
	      "M.12-K.12",
	      "M.18-K.18",
	      "M.24-K.24",
	      "M.30-K.30",
	      "M.36-K.36",
	      "M.48-K.48",
	      "M.60-K.60",
	      "M.72-K.72",
	      "M.120-K.120",
	      "M.168-K.168")
K_weekly_de_analysis =  moanin::weekly_differential_expression_analysis(
     data, splines_model, contrasts,
     use_voom_weights=FALSE)
contrasts = c("M.0-C.0",
	      "M.3-C.3",
	      "M.6-C.6",
	      "M.9-C.9",
	      "M.12-C.12",
	      "M.18-C.18",
	      "M.24-C.24",
	      "M.30-C.30",
	      "M.36-K.36",
	      "M.48-C.48",
	      "M.60-C.60",
	      "M.72-C.72",
	      "M.120-C.120",
	      "M.168-C.168")
C_weekly_de_analysis =  moanin::weekly_differential_expression_analysis(
     data, splines_model, contrasts,
     use_voom_weights=FALSE)
contrasts = c("M.0-VH.0",
	      "M.3-VH.3",
	      "M.6-VH.6",
	      "M.9-VH.9",
	      "M.12-VH.12",
	      "M.18-VH.18",
	      "M.24-VH.24",
	      "M.30-VH.30",
	      "M.36-VH.36",
	      "M.48-VH.48",
	      "M.60-VH.60",
	      "M.72-VH.72",
	      "M.120-VH.120",
	      "M.168-VH.168")
VH_weekly_de_analysis =  moanin::weekly_differential_expression_analysis(
     data, splines_model, contrasts,
     use_voom_weights=FALSE)

``` 


Let's look at the distribution of genes found differentially expressed per
week between control and each of the influenza strains.

```{r fig.width=8, fig.height=8, echo=FALSE}
par(mfrow=c(2, 2))

time = c(0, 3, 6, 9, 12, 18, 24, 30, 36, 48, 60, 72, 120, 168)

timely_barplot = function(de_analysis, labels, qval_threshold=0.05, title=""){
    qval_colnames = colnames(de_analysis)[
	grepl("qval", colnames(de_analysis))]
    number_de_genes_per_time = colSums(de_analysis[, qval_colnames] < 0.05)
    names(number_de_genes_per_time) = labels
    barplot(number_de_genes_per_time, col="black", ylim=c(0, 20000),
	    xlab="Timepoint", ylab="Number of DE genes", main=title)
    
}

timely_barplot(K_weekly_de_analysis, time, title="Kawasaki")
timely_barplot(C_weekly_de_analysis, time, title="California")
timely_barplot(weekly_de_analysis, time, title="Vietnam low-dose")
timely_barplot(VH_weekly_de_analysis, time, title="Vietnam high-dose")
```

The distribution of number of genes found differentially expressed by
considering each time-point independantly highlights the challenges of such
approach. Timepoint 6H and 18H of the Kawasaki strain have fewer numbers of
genes found differentially expressed than timepoints 3H and 9H: this is likely
due to biological or technical variances for specific genes at specific
timepoints.

As a summary, classic differential expression methods are appropriate for
unordered treatments, but fail to use the temporal structure of the data.

### Time-course differential expression analysis between two groups

To leverage this temporal structure, Storey et al [@storey:significance] proposes to
model time-course micro-array with splines functions, and to use a log-ratio
likelihood test to detect differentially expressed genes. `moanin` extends
this idea by providing functions to perform time-course differential
expression analysis using `limma` contrasts.


```{r de_analysis}
# Define contrast
contrasts = limma::makeContrasts(
	contrasts=timecourse_contrasts,
	levels=levels(meta$Group))

# The function takes the data (data.frame or named matrix), the meta data
# (data.frame containing a timepoint and group column, the first corresponding
# to the time-course information, the latter corresponding to the
# treatment).
pvalues = moanin::timecourse_differential_expression_analysis(
    data, splines_model, contrasts)
qvalues = apply(pvalues, 2, stats::p.adjust)

```

The number of genes found differentially expressed ranges from around 12000 to
29000 depending on the strain and dosage of influenza virus given to the mice.
This corresponds to between 30% to 70% of the genes found differentially
expressed in this time-course experiment.

```{r echo=FALSE, fig.width=6, fig.height=6}
num_de = colSums(qvalues < 0.05)
names(num_de) = c(
    "Kawasaki", "California", "Vietnam LD", "Vietname HD")
barplot(num_de, col="black", ylab="Number of DE genes")
```

The next step in a classical differential expression analysis is typically to
assess the effect of the treatment by looking at the log fold change.
Computing the log fold change on a time-course experiment is not trivial: one
can be interested in the average log-fold change across time, or the
cumulative log-fold change. Sometimes a gene can be over-expressed at the
beginning of the time-course data, and then over-expressed at the end of the
experiment. As a result, `moanin` provides a number of possible ways to
compute the log fold change across the whole time-course.

First, `moanin` provides as simple interface to compute the log-fold change
for each individual timepoints.

```{r}
log_fold_change_timepoints = moanin::estimate_log_fold_change(
    data, splines_model, contrasts,  method="timely")
```

```{r echo=FALSE}
kable(head(log_fold_change_timepoints[,
colnames(log_fold_change_timepoints)[1:5]]))
```

This matrix can then be used to visualize the log-fold change for each
contrast per timepoint.

Sometimes, a single value per gene and per contrast is more useful. Here is a
table of the possible ways to compute log-fold change values.

| Name   | Formula   |     |
|------- | --------- |--- |
| timely | $\text{lfc}(t)$  | Function of time |
| sum  | $\sum_t lfc(t)$  | Sum of log fold change. |
| abs_sum  | $\sum_t \|lfc(t)\|$  | Always positive |
| max  | $\max_t \|lfc(t)\|$  | Always positive  |
| min  | $\min_t \|lfc(t)\|$  | Always positive |
| epicon  | $$  | Captures overall strength of response and overall direction |


```{r}
log_fold_change_epicon = moanin::estimate_log_fold_change(
    data, splines_model, contrasts,  method="epicon")

log_fold_change_sum = moanin::estimate_log_fold_change(
    data, splines_model, contrasts,  method="sum")

log_fold_change_max = moanin::estimate_log_fold_change(
    data, splines_model, contrasts, method="max")

log_fold_change_min = moanin::estimate_log_fold_change(
    data, splines_model, contrasts, method="min")
```


```{r echo=FALSE, fig.height=8, fig.width=8, message=FALSE}
par(mfrow=c(2, 2))
plot(log_fold_change_epicon[, "M-C"],
     log_fold_change_sum[, "M-C"], pch=20, main="EPICON vs sum",
     xlab="EPICON", ylab="sum")
plot(log_fold_change_epicon[, "M-C"],
     log_fold_change_max[, "M-C"], pch=20, main="EPICON vs max",
     xlab="EPICON", ylab="max")
plot(log_fold_change_max[, "M-C"],
     log_fold_change_min[, "M-C"], pch=20, main="max vs min",
     xlab="max", ylab="min")
plot(log_fold_change_max[, "M-C"],
     log_fold_change_sum[, "M-C"], pch=20, main="max vs sum",
     xlab="max", ylab="sum")
```

From the single measures of log-fold change and the p-value, we can now look
at the volcano plot. Here is an example of a volcano plot for the comparaison
of the control to the Kawasaki strain, using the EPICON log fold change
computation.

```{r fig.height=8, fig.width=8}
pvalue = qvalues[, "M-K"]
lfc_epicon = log_fold_change_epicon[, "M-K"]
names(lfc_epicon) = row.names(log_fold_change_epicon)

plot(lfc_epicon, -log10(pvalue), pch=20, main="Volcano plot",
     xlim=c(-2.5, 2))
```

As another sanity check, `moanin` provides a simple utility function to
visualize gene time-course data. Here, we plot the 12 genes with the smallest
p-values.

```{r fig.height=8, fig.width=8}
top_DE_genes_pval = names(sort(pvalue)[1:12])
moanin::plot_genes(data[top_DE_genes_pval, ], splines_model,
		   colors=ann_colors$Group, smooth=TRUE)
```

And here we visualize the genes with the largest absolute EPICON log-fold
change.

```{r fig.height=8, fig.width=8}
top_DE_genes_lfc = names(
    sort(abs(lfc_epicon),
	 decreasing=TRUE)[1:12])
moanin::plot_genes(data[top_DE_genes_lfc, ], splines_model,
		   colors=ann_colors$Group, smooth=TRUE)

```

Thanks to those visualization, we can see that genes often follow similar
patterns of expression, although on a different scale for each gene. We can
leverage this observation to cluster the genes into groups of similar patterns
of transcriptomic response.

```{r echo=FALSE, results="hide"}
# Rename stuff for saving
outname = ".results/pvalues.txt"
colnames(pvalues) = unlist(
    lapply(colnames(pvalues), function(x){paste0(x, "-pval")}))
    
colnames(qvalues) = unlist(
    lapply(colnames(qvalues), function(x){paste0(x, "-qval")}))

colnames(log_fold_change_epicon) = unlist(
    lapply(colnames(log_fold_change_epicon),
    function(x){paste0(x, "-lfc_epicon")}))

colnames(log_fold_change_max) = unlist(
    lapply(colnames(log_fold_change_max),
    function(x){paste0(x, "-lfc_max")}))

de_analysis = merge(pvalues, qvalues, all=TRUE, by="row.names")
row.names(de_analysis) = row.names(pvalues)
de_analysis = subset(de_analysis, select=-c(Row.names))

de_analysis = merge(de_analysis, log_fold_change_max, all=TRUE, by="row.names")
row.names(de_analysis) = row.names(pvalues)
de_analysis = subset(de_analysis, select=-c(Row.names))

de_analysis = merge(de_analysis, log_fold_change_epicon, all=TRUE,
		    by="row.names")
row.names(de_analysis) = row.names(pvalues)
de_analysis = subset(de_analysis, select=-c(Row.names))

write.table(de_analysis, outname, sep="\t")
```


# References
