---
title: "Normalization: A pipeline to analyse time-course gene expression data"
output: rmarkdown::html_vignette
---

## Differential expression analysis of time-course data

```{r echo=FALSE, results="hide"}
# Reloading necessary data and preliminary results
library(devtools)
library(limma)
library(splines)
library(stats)
library(moanin)
```

```{r echo=FALSE}
# Set options
df = 6

# Preprocessing & filtering
is_count = FALSE
take_log = FALSE    
filter_expression = FALSE

# Differential expression analysis
timecourse_contrasts = c("C-K", "C-M")

# Clustering
filter_genes = TRUE
n_genes_to_keep = 5000
n_clusters = 20
percentage_genes_to_label = 0.5
```


```{r echo=FALSE, results="hide"}
# Set data files and options
data("shoemaker2015")
data = shoemaker2015$data
meta = shoemaker2015$meta

# Right now, the package doesn't contain the normalized data
```

### Time-course differential expression analysis between two groups

FIXME add estimation of log fold change.

```{r de_analysis}
# Define contrast
de_analysis = data.frame(row.names=row.names(data))
for(contrast_formula in timecourse_contrasts){
    contrast = limma::makeContrasts(
	contrast_formula,
	levels=levels(meta$Group))
    model = moanin:::edgeWithContrasts(as.matrix(data), meta, contrast)
    contrast_name = gsub(" ", "", contrast_formula, fixed=TRUE)
    contrast_name_pval = paste(contrast_name, "-pval", sep="")
    contrast_name_qval = paste(contrast_name, "-qval", sep="")
    de_analysis[contrast_name_pval] = model$pval_lrt
    de_analysis[contrast_name_qval] = model$qval_lrt

    # Now compute log fold change using different methods
    # XXX This fails very badly if the contrasts provided is wrong.
    log_fold_change_epicon = moanin::estimate_log_fold_change(
	data, meta, c(contrast_formula), method="epicon")
    contrast_name_lfc_epicon = paste(contrast_name, "-lfc_epicon", sep="")
    de_analysis[contrast_name_lfc_epicon] = log_fold_change_epicon

    # XXX Something is wrong with method max
    log_fold_change_max = moanin::estimate_log_fold_change(
	data, meta, c(contrast_formula), method="max")
    contrast_name_lfc_max = paste(contrast_name, "-lfc_max", sep="")
    de_analysis[contrast_name_lfc_max] = log_fold_change_max

}

pvalues = de_analysis
outname = ".results/pvalues.txt"
write.table(pvalues, outname, sep="\t")
```


### Developmental differential expression analysis

```{r}
# Define contrast
contrast_formula = "C"
contrast = limma::makeContrasts(
    contrast_formula,
    levels=levels(meta$Group))
# Why do I have to do this again?
mask = with(meta, Group == contrast_formula)
model = moanin:::edgeWithContrasts(
    as.matrix(data), meta, contrast, developmental=TRUE,
    mask=mask)
```

### Weekly differential expression analysis

```{r}
# Define contrast
contrasts = c("C.0-K.0",
	      "C.3-K.3",
	      "C.6-K.6",
	      "C.9-K.9",
	      "C.12-K.12",
	      "C.18-K.18",
	      "C.24-K.24",
	      "C.30-K.30",
	      "C.48-K.48",
	      "C.60-K.60",
	      "C.120-K.120",
	      "C.168-K.168")
	
meta$WeeklyGroup = make.names(meta$Group:as.factor(meta$Time))
# de_analysis =  fit_weekly_analysis(
#     data, meta, contrasts,
#     use_voom_weights=FALSE)
```
