---
title: "Downstream analysis: A pipeline to analyse time-course gene expression data"
output: rmarkdown::html_vignette
---


```{r echo=FALSE, results="hide"}
# Loading dependencies
library(devtools)
library(limma)
library(splines)
library(stats)
library(moanin)
library(KEGGprofile)

# library(moanin)
```

```{r echo=FALSE, results="hide"}
# Set options
df = 6

# Preprocessing & filtering
is_count = FALSE
take_log = FALSE    
filter_expression = FALSE

# Differential expression analysis
timecourse_contrasts = c("C-K", "C-M")

# Clustering
filter_genes = TRUE
n_genes_to_keep = 5000
n_clusters = 20
percentage_genes_to_label = 0.5
```

```{r echo=FALSE, results="hide"}
# Set data files and options
data("shoemaker2015")
data = shoemaker2015$data
meta = shoemaker2015$meta

labels = read.table(".results/clustering_labels.txt", sep="\t",
		    check.names=FALSE)
```

```{r}
gene_names = row.names(labels)
gene_names = gene_names[!is.na(labels)]
labels = labels[!is.na(labels)]
```

### Finding enriched pathways


```{r}
library(biomaRt)
ensembl = useMart("ensembl")
ensembl = useDataset("mmusculus_gene_ensembl", mart=ensembl)

```

```{r}
clusters = as.vector(unlist(unique(labels)))
for(cluster in clusters){
    print(cluster)
    genes = gene_names[labels == cluster]

    # convert gene names
    genes = getBM(attributes=c("ensembl_gene_id", "entrezgene"),
		       filters="refseq_mrna", values=genes,
		       mart=ensembl)["entrezgene"]
    genes = as.vector(unlist(genes))
    pathways = KEGGprofile::find_enriched_pathway(
	genes, species="mmu",
	download_latest=TRUE)
    print(pathways$stastic)
}
```


### Finding enriched GO terms
