---
title: "Conclusion: A pipeline to analyse time-course gene expression data"
output: rmarkdown::html_vignette
---

# Conclusion

This workflow provides a tutoral for the analysis of time-course gene
expression data in R, illustrated through the analysis of mice lung tissue
exposed to different influenza strain. It covers four main steps; (1) quality
control and normalization; (2) differential expression analysis; (3)
clustering of time-course gene expression data; (4) downstream analysis of
clusters.

# Software and data availability

The source code for this workflow can be found at **FIXME** 
Data used in this workflow are available from NCBI GEO, accession GSE95601.

# Author contributions

NV and EP wrote the workflow.

# Competing interests
The authors declare that they have no competing interests.

# Grant information

This research was funded in part by a Department of Energy (DOE) grant
(DE-SC0014081); by the Gordon and Betty Moore Foundation (Grant GBMF3834) and
the Alfred P. Sloan Foundation (Grant 2013-10-27) to the University of
California, Berkeley [N.V.]; by a ENS-CFM Data Science Chair [E.P.].

*I confirm that the funders had no role in study design, data collection and
analysis, decision to publish, or preparation of the manuscript.*

# Acknowledgments

The authors thank Karthik Ram and the Ropensci community for valuable
feedback.
