---
title: "A pipeline to analyse time-course gene expression data"
author:
    - name: Nelle Varoquaux
      affiliation: Department of Statistics and Berkeley Institute for Data Science, University of California, Berkeley, Berkeley, CA 
    - name: Elizabeth Purdom
      affiliation: Department of Statistics, University of California, Berkeley, Berkeley, CA

abstract: >
    The phenotypic diversity of cells is governed by a complex equilibrium
    between their genetic identity and their environmental interactions:
    Understanding the dynamics of gene expression is a fundamental question of
    biology. However, analysing time-course transcriptomic data raises unique
    challenging statistical and computational questions, requiring the
    development of novel methods and software.  Using as case study
    time-course transcriptomics data from mice exposed to different strains of
    influenza, this workflows provides a step-by-step tutorial of the
    methodology used to analyse time-course data: (1) normalization of the
    micro-array dataset; (2) differential expression analysis using functional
    data analysis; (3) clustering fo time-course data; (4) interpreting
    clusters with GO term and KEGG pathway enrichment analysis.

keywords: >
    time-course gene expression data, clustering, differential expression,
    workflow
vignette: >
    %\VignetteIndexEntry{A worfklow for low-level analyses of scRNA-seq data}
    %\VignetteEngine{knitr::rmarkdown}
#output: BiocWorkflowTools::f1000_article
output: pdf_document

bibliography: bibliography.bib
---

```{r echo=FALSE, results="hide", message=FALSE}
# Loading dependencies
library(devtools)
library(limma)
library(splines)
library(stats)
library(knitr)
library(kableExtra)
#library(BiocStyle)
library(biomaRt)

library(ggplot2)
library(NMF)
library(KEGGprofile)
library(viridis)

library(topGO)

library(moanin)

source("utils.R")
```

```{r echo=FALSE, include=FALSE}
knitr::opts_chunk$set(
  cache=TRUE, autodep=TRUE, warning=FALSE, error=FALSE, message=FALSE,
  echo=TRUE,
  duplicate.label="allow"
)
```


# TODO list


- Add legends to the plots that don't have.
- Currently have two functions "plot_genes", "plot_centroids", that do exactly
  the same thing. Find a better name for both of them


# Introduction

Gene expression studies provide simultaneous quantification of the level of
mRNA from all genes in a sample. High-throughput studies of gene expression
have a long history, starting with microarray technologies in the 1990s
through to single-cell technologies. While many expression studies are
designed to compare the gene expression in distinct groups, there is also a
long history of time-course expression studies, where the the gene expression
is compared across time by measuring mRNA levels from different samples across
time [^1]. Such time course studies can vary from measuring a few distinct time
points, to sampling ten to twenty time points. Many longer time series are
particularly interested in investigating development over time. More recently,
single-cell studies track single cells through their development, and a single
cell is measured at a particular moment in its developmental progression -- a
value that is not know but estimated from the data as its "pseudo-time."

[^1]: Because the collection of the mRNA is often destructive, samples at
different time points are generally from different biological samples;
longitudinal studies, for example tracking the same subject over time, are
certainly possible in certain settings, but not directly considered here. 

While there are many methods that have been proposed for discrete aspects of
time course data, the entire workflow for analysis of such data remains
difficult, particularly for long, developmental time series. Most methods
proposed for time course data are concerned with detecting genes that are
changing over time (differential expression analysis), examples being `edge`
[@storey:significance], functional component analysis based models [@wu:more],
time-course permutation tests [@park:statistical], and multiple testing
strategies to combine singe time point differential expression analysis
[@sun:multiple]. However, with long time course datasets, particularly in
developmental systems, a massive number of genes will show *some* change. For
example, in the mice lung tissues infected with influenza, over 50% of genes
are shown to be changing over time. The task in these settings is often not to
detect changes in genes, but to categorize into biologically interpretable
patterns the vast number of changes discovered. 

We present here a workflow for such an analysis that consists of 4 main
parts (Figure \ref{fig:workflow}):

- Quality control and normalization;
- Identification of genes that are differentially expressed;
- Clustering of genes into distinct temporal patterns;
- Biological interpretation of the clusters.

This workflow represents an integration of both novel implementations of
previously established methods and new methodologies for the settings of
developmental time series. It relies on several standard packages for
analysing gene expression data, some specific for time-course data, others
broadly used by the community. We provide the various steps of the workflow as
functions in a R package called `moanin`.

```{r schema, echo=FALSE, fig.width=4, fig.cap="Workflow for analyzing time-course datasets."}

knitr::include_graphics("images/fig1_schemas.png")

```


